muffnn/autoencoder/autoencoder.py

```diff
@@ -0,0 +1,686 @@
+                learning_rate=self.learning_rate).minimize(self._obj_func)
+
+    def _build_output_layer_and_scores(self, t):
+        """Add ops for output layer and scores to the graph."""
+        scores = 0.0
+
+        # Below a transpose on `t` is needed if the TF op `gather` is used
+        # on the features dimension of `t` (i.e., dim 2).
+        # A shortcut has been added here to avoid a transpose if it is not
+        # needed.
+        if (self._categorical_indices is None and
+                self._binary_indices is None):
+
+            if self.metric == 'mse':
+                if self.output_activation is not None:
+                    t = self.output_activation(t)
+
+                diff = t - self._input_values
+                scores = tf.reduce_sum(tf.square(diff),
+                                       reduction_indices=[1])
+            elif self.metric == 'cross-entropy':
+                if self.output_activation is tf.nn.sigmoid:
+                    scores += tf.reduce_sum(
+                        tf.nn.sigmoid_cross_entropy_with_logits(
+                            t, self._input_values),
+                        reduction_indices=[1])
+                    t = tf.nn.sigmoid(t)
+                elif self.output_activation is tf.nn.softmax:
+                    scores += tf.nn.softmax_cross_entropy_with_logits(
+                        t, self._input_values, dim=1)
+                    t = tf.nn.softmax(t)
+                else:
+                    raise ValueError("Only `tensorflow.nn.sigmoid` and "
+                                     "`tensorflow.nn.softmax` output "
+                                     "activations can be used with the "
+                                     "'cross-entropy' metric!")
+            else:
+                raise ValueError('Metric "%s" is not allowed!' % self.metric)
+        else:
+            # Use the transpose since gather works on first dim.
+            t = tf.transpose(t)
+
+            # Categorical vars w/ one-hot encoding.
+            # Softmax all of the categorical stuff and use cross-entropy.
+            if self._categorical_indices is not None:
+                min_float32 = tf.constant(tf.float32.min)
+                for i, begin, size in zip(
+                        range(self._categorical_indices.shape[0]),
+                        self._categorical_indices[:, 0],
+                        self._categorical_indices[:, 1]):
+                    scores += tf.nn.softmax_cross_entropy_with_logits(
+                        tf.slice(t, [begin, 0], [size, -1]),
+                        tf.slice(tf.transpose(self._input_values),
+                                 [begin, 0], [size, -1]),
+                        dim=0)
+
+                    # This one is painful. TensorFlow does not, AFAIK,
+                    # support assignments to Tensors that come out of
+                    # operations (it does support assignments to Variables).
```

Comment:
The case condition in `tf.case` is a boolean scalar, so this won't get you around the mask thing.